File: api/src/main/java/com/api/app/service/basket/BasketServiceImpl.java
package com.api.app.service.basket;

import com.api.app.common.exception.ApiError;
import com.api.app.common.exception.ApiException;
import com.api.app.dto.request.basket.BasketAddRequest;
import com.api.app.dto.request.basket.BasketModifyRequest;
import com.api.app.dto.response.basket.BasketResponse;
import com.api.app.entity.BasketBase;
import com.api.app.entity.MemberBase;
import com.api.app.repository.basket.BasketBaseMapper;
import com.api.app.repository.basket.BasketBaseTrxMapper;
import com.api.app.repository.member.MemberBaseMapper;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

/**
 * 장바구니 서비스 구현
 *
 * @author system
 * @version 1.0
 * @since 2025-10-29
 */
@Slf4j
@Service
@RequiredArgsConstructor
public class BasketServiceImpl implements BasketService {

    private final BasketBaseMapper basketBaseMapper;
    private final BasketBaseTrxMapper basketBaseTrxMapper;
    private final MemberBaseMapper memberBaseMapper;

    @Override
    public List<BasketResponse> getBasketList(String email) {
        log.debug("장바구니 목록 조회 시작: email={}", email);

        // 회원번호 조회
        String memberNo = getMemberNoByEmail(email);

        // 장바구니 목록 조회
        List<BasketResponse> basketList = basketBaseMapper.selectBasketListByMemberNo(memberNo);

        log.debug("장바구니 목록 조회 완료: email={}, count={}", email, basketList.size());
        return basketList;
    }

    @Override
    @Transactional
    public String addBasket(String email, BasketAddRequest request) {
        log.debug("장바구니 추가 시작: email={}, goodsNo={}, itemNo={}",
                email, request.getGoodsNo(), request.getItemNo());

        // 회원번호 조회
        String memberNo = getMemberNoByEmail(email);

        // 동일한 상품/단품이 이미 있는지 확인
        BasketBase existingBasket = basketBaseMapper.selectBasketBaseByMemberAndItem(
                memberNo, request.getGoodsNo(), request.getItemNo());

        String basketNo;

        if (existingBasket != null) {
            // 이미 있으면 수량 증가
            existingBasket.setQuantity(existingBasket.getQuantity() + request.getQuantity());
            existingBasket.setModifyId(memberNo);

            int result = basketBaseTrxMapper.updateBasketBase(existingBasket);
            if (result != 1) {
                throw new ApiException(ApiError.INTERNAL_SERVER_ERROR, "장바구니 수량 수정에 실패했습니다");
            }

            basketNo = existingBasket.getBasketNo();
            log.info("장바구니 수량 증가 완료: basketNo={}, newQuantity={}",
                    basketNo, existingBasket.getQuantity());

        } else {
            // 없으면 새로 추가
            basketNo = basketBaseTrxMapper.generateBasketNo();
            log.debug("장바구니번호 생성: {}", basketNo);

            BasketBase basketBase = new BasketBase();
            basketBase.setBasketNo(basketNo);
            basketBase.setMemberNo(memberNo);
            basketBase.setGoodsNo(request.getGoodsNo());
            basketBase.setItemNo(request.getItemNo());
            basketBase.setQuantity(request.getQuantity());
            basketBase.setIsOrder(false);
            basketBase.setRegistId(memberNo);

            int result = basketBaseTrxMapper.insertBasketBase(basketBase);
            if (result != 1) {
                throw new ApiException(ApiError.INTERNAL_SERVER_ERROR, "장바구니 추가에 실패했습니다");
            }

            log.info("장바구니 추가 완료: basketNo={}, goodsNo={}, itemNo={}",
                    basketNo, request.getGoodsNo(), request.getItemNo());
        }

        return basketNo;
    }

    @Override
    @Transactional
    public void modifyBasket(String email, String basketNo, BasketModifyRequest request) {
        log.debug("장바구니 수정 시작: email={}, basketNo={}", email, basketNo);

        // 회원번호 조회
        String memberNo = getMemberNoByEmail(email);

        // 장바구니 조회 및 권한 확인
        BasketBase basketBase = basketBaseMapper.selectBasketBaseByBasketNo(basketNo);
        if (basketBase == null) {
            throw new ApiException(ApiError.DATA_NOT_FOUND, "장바구니를 찾을 수 없습니다");
        }

        if (!basketBase.getMemberNo().equals(memberNo)) {
            throw new ApiException(ApiError.FORBIDDEN, "본인의 장바구니만 수정할 수 있습니다");
        }

        // 수정할 필드 적용
        if (request.getGoodsNo() != null) {
            basketBase.setGoodsNo(request.getGoodsNo());
        }
        if (request.getItemNo() != null) {
            basketBase.setItemNo(request.getItemNo());
        }
        if (request.getQuantity() != null) {
            basketBase.setQuantity(request.getQuantity());
        }
        basketBase.setModifyId(memberNo);

        int result = basketBaseTrxMapper.updateBasketBase(basketBase);
        if (result != 1) {
            throw new ApiException(ApiError.INTERNAL_SERVER_ERROR, "장바구니 수정에 실패했습니다");
        }

        log.info("장바구니 수정 완료: basketNo={}", basketNo);
    }

    @Override
    @Transactional
    public void deleteBasket(String email, String basketNo) {
        log.debug("장바구니 삭제 시작: email={}, basketNo={}", email, basketNo);

        // 회원번호 조회
        String memberNo = getMemberNoByEmail(email);

        // 장바구니 조회 및 권한 확인
        BasketBase basketBase = basketBaseMapper.selectBasketBaseByBasketNo(basketNo);
        if (basketBase == null) {
            throw new ApiException(ApiError.DATA_NOT_FOUND, "장바구니를 찾을 수 없습니다");
        }

        if (!basketBase.getMemberNo().equals(memberNo)) {
            throw new ApiException(ApiError.FORBIDDEN, "본인의 장바구니만 삭제할 수 있습니다");
        }

        // 장바구니 삭제
        int result = basketBaseTrxMapper.deleteBasketBase(basketNo);
        if (result != 1) {
            throw new ApiException(ApiError.INTERNAL_SERVER_ERROR, "장바구니 삭제에 실패했습니다");
        }

        log.info("장바구니 삭제 완료: basketNo={}", basketNo);
    }

    @Override
    @Transactional
    public void deleteBaskets(String email, List<String> basketNos) {
        log.debug("장바구니 여러 개 삭제 시작: email={}, count={}", email, basketNos.size());

        if (basketNos == null || basketNos.isEmpty()) {
            throw new ApiException(ApiError.INVALID_PARAMETER, "삭제할 장바구니를 선택해주세요");
        }

        // 회원번호 조회
        String memberNo = getMemberNoByEmail(email);

        // 각 장바구니의 권한 확인
        for (String basketNo : basketNos) {
            BasketBase basketBase = basketBaseMapper.selectBasketBaseByBasketNo(basketNo);
            if (basketBase == null) {
                throw new ApiException(ApiError.DATA_NOT_FOUND,
                        String.format("장바구니를 찾을 수 없습니다: %s", basketNo));
            }

            if (!basketBase.getMemberNo().equals(memberNo)) {
                throw new ApiException(ApiError.FORBIDDEN, "본인의 장바구니만 삭제할 수 있습니다");
            }
        }

        // 장바구니 일괄 삭제
        int result = basketBaseTrxMapper.deleteBasketBaseByNos(basketNos);
        log.info("장바구니 여러 개 삭제 완료: email={}, deleted={}", email, result);
    }

    @Override
    @Transactional
    public void deleteAllBaskets(String email) {
        log.debug("장바구니 전체 삭제 시작: email={}", email);

        // 회원번호 조회
        String memberNo = getMemberNoByEmail(email);

        // 장바구니 전체 삭제
        int result = basketBaseTrxMapper.deleteBasketBaseByMemberNo(memberNo);
        log.info("장바구니 전체 삭제 완료: email={}, deleted={}", email, result);
    }

    /**
     * 이메일로 회원번호 조회
     *
     * @param email 이메일
     * @return 회원번호
     */
    private String getMemberNoByEmail(String email) {
        MemberBase member = memberBaseMapper.selectMemberBaseByEmail(email);
        if (member == null) {
            throw new ApiException(ApiError.DATA_NOT_FOUND, "회원 정보를 찾을 수 없습니다");
        }
        return member.getMemberNo();
    }
}

File: api/src/main/kotlin/com/api/app/service/basket/BasketServiceImpl.kt
package com.api.app.service.basket

import com.api.app.common.exception.ApiError
import com.api.app.common.exception.ApiException
import com.api.app.dto.request.basket.BasketAddRequest
import com.api.app.dto.request.basket.BasketModifyRequest
import com.api.app.dto.response.basket.BasketResponse
import com.api.app.entity.BasketBase
import com.api.app.repository.rodb.basket.BasketBaseRepository
import com.api.app.repository.rodb.basket.BasketProjection
import com.api.app.repository.rodb.member.MemberBaseRepository
import com.api.app.repository.rwdb.basket.BasketBaseTrxRepository
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional(readOnly = true)
class BasketServiceImpl(
    private val basketBaseRepository: BasketBaseRepository,
    private val basketBaseTrxRepository: BasketBaseTrxRepository,
    private val memberBaseRepository: MemberBaseRepository
) : BasketService {

    private val log = LoggerFactory.getLogger(this::class.java)

    override fun getBasketList(email: String): List<BasketResponse> {
        val memberNo = getMemberNoByEmail(email)
        return basketBaseRepository.selectBasketListByMemberNo(memberNo).map { it.toResponse() }
    }

    @Transactional
    override fun addBasket(email: String, request: BasketAddRequest): String {
        log.debug("장바구니 추가 시작: email={}, goodsNo={}, itemNo={}", email, request.goodsNo, request.itemNo)

        val memberNo = getMemberNoByEmail(email)

        val existing = basketBaseRepository.findByMemberNoAndGoodsNoAndItemNoAndIsOrderFalse(
            memberNo, request.goodsNo, request.itemNo
        )

        return if (existing != null) {
            existing.quantity += request.quantity
            basketBaseTrxRepository.save(existing)
            log.info("장바구니 수량 증가 완료: basketNo={}", existing.basketNo)
            existing.basketNo
        } else {
            val basket = BasketBase().apply {
                this.memberNo = memberNo
                this.goodsNo = request.goodsNo
                this.itemNo = request.itemNo
                this.quantity = request.quantity
                this.isOrder = false
            }
            basketBaseTrxRepository.save(basket)
            log.info("장바구니 추가 완료: basketNo={}", basket.basketNo)
            basket.basketNo
        }
    }

    @Transactional
    override fun modifyBasket(email: String, basketNo: String, request: BasketModifyRequest) {
        val memberNo = getMemberNoByEmail(email)

        val basket = basketBaseRepository.findById(basketNo).orElseThrow {
            ApiException(ApiError.DATA_NOT_FOUND, "장바구니를 찾을 수 없습니다")
        }

        if (basket.memberNo != memberNo) {
            throw ApiException(ApiError.FORBIDDEN, "본인의 장바구니만 수정할 수 있습니다")
        }

        request.goodsNo?.let { basket.goodsNo = it }
        request.itemNo?.let { basket.itemNo = it }
        request.quantity?.let { basket.quantity = it }

        basketBaseTrxRepository.save(basket)
        log.info("장바구니 수정 완료: basketNo={}", basketNo)
    }

    @Transactional
    override fun deleteBasket(email: String, basketNo: String) {
        val memberNo = getMemberNoByEmail(email)

        val basket = basketBaseRepository.findById(basketNo).orElseThrow {
            ApiException(ApiError.DATA_NOT_FOUND, "장바구니를 찾을 수 없습니다")
        }

        if (basket.memberNo != memberNo) {
            throw ApiException(ApiError.FORBIDDEN, "본인의 장바구니만 삭제할 수 있습니다")
        }

        basketBaseTrxRepository.deleteById(basketNo)
        log.info("장바구니 삭제 완료: basketNo={}", basketNo)
    }

    @Transactional
    override fun deleteBaskets(email: String, basketNos: List<String>) {
        if (basketNos.isEmpty()) throw ApiException(ApiError.INVALID_PARAMETER, "삭제할 장바구니를 선택해주세요")

        val memberNo = getMemberNoByEmail(email)

        basketNos.forEach { basketNo ->
            val basket = basketBaseRepository.findById(basketNo).orElseThrow {
                ApiException(ApiError.DATA_NOT_FOUND, "장바구니를 찾을 수 없습니다: $basketNo")
            }
            if (basket.memberNo != memberNo) {
                throw ApiException(ApiError.FORBIDDEN, "본인의 장바구니만 삭제할 수 있습니다")
            }
        }

        basketBaseTrxRepository.deleteByBasketNoIn(basketNos)
        log.info("장바구니 여러 개 삭제 완료: email={}, count={}", email, basketNos.size)
    }

    @Transactional
    override fun deleteAllBaskets(email: String) {
        val memberNo = getMemberNoByEmail(email)
        basketBaseTrxRepository.deleteByMemberNo(memberNo)
        log.info("장바구니 전체 삭제 완료: email={}", email)
    }

    private fun getMemberNoByEmail(email: String): String =
        memberBaseRepository.findByEmail(email)?.memberNo
            ?: throw ApiException(ApiError.DATA_NOT_FOUND, "회원 정보를 찾을 수 없습니다")

    private fun BasketProjection.toResponse() = BasketResponse(
        basketNo = getBasketNo(),
        memberNo = getMemberNo(),
        goodsNo = getGoodsNo(),
        goodsName = getGoodsName(),
        goodsStatusCode = getGoodsStatusCode(),
        goodsMainImageUrl = getGoodsMainImageUrl(),
        salePrice = getSalePrice(),
        itemNo = getItemNo(),
        itemName = getItemName(),
        itemPrice = getItemPrice(),
        itemStatusCode = getItemStatusCode(),
        stock = getStock(),
        quantity = getQuantity(),
        isOrder = getIsOrder(),
        registDateTime = getRegistDateTime()
    )
}
